unified-scheduler-pool/src/lib.rs

```diff
@@ -194,34 +199,535 @@ impl TaskHandler for DefaultTaskHandler {
+        executed_task: &mut Box<ExecutedTask>,
+        handler_context: &HandlerContext,
+    ) {
+        debug!("handling task at {:?}", thread::current());
+        TH::handle(
+            &mut executed_task.result_with_timings.0,
+            &mut executed_task.result_with_timings.1,
+            bank,
+            executed_task.task.transaction(),
+            executed_task.task.task_index(),
+            handler_context,
+        );
+    }
+
+    fn take_session_result_with_timings(&mut self) -> ResultWithTimings {
+        self.session_result_with_timings.take().unwrap()
+    }
+
+    fn put_session_result_with_timings(&mut self, result_with_timings: ResultWithTimings) {
+        assert_matches!(
+            self.session_result_with_timings
+                .replace(result_with_timings),
+            None
+        );
+    }
+
+    fn start_threads(&mut self, context: &SchedulingContext) {
+        let (mut runnable_task_sender, runnable_task_receiver) =
+            chained_channel::unbounded::<Task, SchedulingContext>(context.clone());
+        let (executed_task_sender, executed_task_receiver) = unbounded::<ExecutedTaskPayload>();
+        let (finished_task_sender, finished_task_receiver) = unbounded::<Box<ExecutedTask>>();
+        let (accumulated_result_sender, accumulated_result_receiver) =
+            unbounded::<Option<ResultWithTimings>>();
+
+        let mut result_with_timings = self.session_result_with_timings.take();
+
+        // High-level flow of new tasks:
+        // 1. the replay stage thread send a new task.
+        // 2. the scheduler thread accepts the task.
+        // 3. the scheduler thread dispatches the task after proper locking.
+        // 4. the handler thread processes the dispatched task.
+        // 5. the handler thread reply back to the scheduler thread as an executed task.
+        // 6. the scheduler thread post-processes the executed task.
+        // 7. the scheduler thread send the executed task to the accumulator thread.
+        // 8. the accumulator thread examines the executed task's result and accumulate its timing,
+        //    finally dropping the transaction inside the executed task.
+        let scheduler_main_loop = || {
+            let handler_count = self.handler_count;
+            let session_result_sender = self.session_result_sender.clone();
+            let new_task_receiver = self.new_task_receiver.clone();
+
+            let mut session_ending = false;
+            let mut active_task_count: usize = 0;
+
+            // Now, this is the main loop for the scheduler thread, which is a special beast.
+            //
+            // That's because it's the most notable bottleneck of throughput. Unified scheduler's
+            // overall throughput is largely dependant on its ultra-low latency characteristic,
+            // which is the most important design goal of the scheduler in order to reduce the
+            // transaction confirmation latency for end users.
```

Comment:
> It might be the bottleneck of just scheduling, but the main bottleneck of replay will still be the processing of transactions (load, execute, commit), right?

Currently, yes. given ~100 handler threads, scheduler thread can be the bottleneck. and I'm developing this scheduler with that future in mind.

> Correct me if I'm wrong, but it seems the sensitiviy is there because between the time of ExecutedTask(Tx0) and Task(Tx1), the handler thread is idle.
> Even with the additional thread which does the accumulation and dropping, there is still going to be some amount of idle time here.

yeah. these are correct.

> Just as a note, I also ran into this in the block-production scheduler in core, but opted to allow the queueing of txs for specific handlers. Even trying to keep just 2 txs always enqueued gives significant leeway in the latency, since as long as the scheduler can queue another in the time it takes to process a transaction there is nearly no idling time.

hmm, this seems odd. busy looping at the other hand of mpmc channels usually gives round-trip less-than-500ns latency, which i'm planning to employ for unified scheduler. i think those threads were yielding to the kernel (ie parking)?

as you pointed out, i can't eliminate the idle time. but then, 1us (500ns for channel + 500ns for scheduling) per a single conflicting tx is acceptable latency.

actually, i wanted to show this as some bench, but i'm running out of time.. i spend too much on the other concern... lol https://github.com/solana-labs/solana/pull/34676#discussion_r1450581025

Reply:
> i think those threads were yielding to the kernel (ie parking)?

possible.

> as you pointed out, i can't eliminate the idle time. but then, 1us (500ns for channel + 500ns for scheduling) per a single conflicting tx is acceptable latency.

what's the added latency from dropping the txs? or in other words, what would be the affect on overall throughput if we did not have the separate "drop" thread? My main concern here was just that we're spawning this new thread as part of the scheduler, and if that's necessary we should be able to back that decision up with some numbers on how much it helps (or how much we are hurt by **not** doing that).

> as you pointed out, i can't eliminate the idle time.

yeah I think without allowing conflicts to actually queue up, this cannot be resolved entirely, just mitigated.

Reply:
thanks for detailing your concern.

i've finally wrote benches: 6120684e5933a908018cde4d4ae930ea5448de88 (won't push to this pr)....:

```
test bench_tx_throughput_drop_in_scheduler ... bench:   8,556,627 ns/iter (+/- 3,168,724)
test bench_tx_throughput_drop_in_accumulator ... bench:   7,334,701 ns/iter (+/- 2,600,740)
test bench_tx_throughput_drop_in_scheduler_conflicting ... bench:  11,403,197 ns/iter (+/- 3,062,017)
test bench_tx_throughput_drop_in_accumulator_conflicting ... bench:  10,794,434 ns/iter (+/- 6,808,692)
```

so, processing 10,000 txes takes always longer when we drop txes in the scheduler thread, regardless conflicting or not.

note that you have to ensure to rebuild when repoto these results.. cargo usually skips rebuild with feature unification...

Reply:
The raw numbers looks like dropping in scheduler takes longer, which isn't surprising since its doing more work; but they are well within the statistical noise. 

`bench_tx_throughput_drop_in_scheduler_conflicting` raw shows 609us higher than `bench_tx_throughput_drop_in_accumulator_conflicting`, but the error bounds are 3ms and 6ms. 

This seems to imply that there's *not* a statistically significant effect from having the separate accumulator thread. Am I missing something in these results?

Reply:
I appreciate the feedback. Seems casual benchmark wasn't enough.

I reverted the accumulator thread (https://github.com/solana-labs/solana/pull/34676/commits/2f592f5a84dba4fa61f92e1c23d1923c7bfea238) for now to land this pr.

Now that I've done somewhat extended synthesized bench-marking on this pr. That said, I still think the accumulator thread performs better in the end.

However, because this pr doesn't contain any actual scheduling impl, i concluded that presenting some convincing numbers is rather difficult. ;) So, I'll try to ship the scheduling impl first, instead of finishing up the scheduler pool (was original plan). Even later, I'll try to land the accumulator thread along with other missing things in the scheduler pool with more proper justification / more realistic benchmark numbers. 

In other words, landing the accumulator thread wasn't essential for now. So, I was too early to ship it.

> Am I missing something in these results?

yea. blame me for not explaining enough. didn't expect the accumulator thread to be so hot...

1. the accumulator thread will contain other offloaded tasks.

Namely, it's used for dumping all detailed timings for individual transactions to influxdb at background fashion (the code is in the base pr). Also, I'm planning to move rpc related blockstore writes here, as this isn't essential for bank hash. These aren't happening in this pr, so the accumulator thread itself was just ahead of it.

2. dropping transactions can be really heavy

in the intention of a quick sanity check bench, i wasn't bothered to do so. but transactions can be really large with many instructions with each having instruction data blobs, pubkeys...

when blocks contain these transactions (it does usually), it clearly throttles the scheduler thread when dropped inside that.

3. medians inside overlapping deviation range don't necessarily mean statistical noise.

firstly, these multi-threaded benches tend to contain high variance in runtime duration. so, i did little attention to the deviations. I just shared the numbers after some casual multiple runs, thinking these numbers are relatively stable enough.


maybe, seeing is believing. :) here's comparable criterion bench results, which i wrote with different parameters/settings:

![image](https://github.com/solana-labs/solana/assets/117807/1df5d69f-38b8-439a-9b1f-748d24024bc9)

![image](https://github.com/solana-labs/solana/assets/117807/175145f1-287e-49ab-b6ae-ce2e76ad7ed8)

note that I think these periodic outliers are due to cpu freq scaling. Sans these outlifers, I think the primary convex of pdf is shifted in favor of dropping in accumlator, clearly showing statistically significant improvement. Also, this is visibly confirmed by plotted individual samples.

> I think we are not really seeing eye-to-eye on this additional thread, and might just be we have different perspectives.

i think we have a constructive discussion. thanks for patience and expressing your perspective.

> Feel free to pull another core contributor in for their opinion on this aspect of the PR.

thanks for suggestion. yeah, we could seek other's opinions, should we be stuck in other topics.

> I don't recall seeing anything else that would block this PR for me, but will take another look once the thread-issue is settled.

that's good to hear. :)